### ISSUE TITLE
Unnecessary copying of data to hepevt record in Fortran

### ISSUE THREAD

**Issue by HDembinski** [impy-project/chromo#50]
I noticed that most models contain a custom piece of that copies data from an internal stack into the standard hepevt record. Since we care about performance, we should avoid those copies, which are not needed. There are only a few that cannot be avoided, like converting the internal ids to pdg ids.

For example in Sibyll2.1 we have the internal block:
```fortran
      COMMON /S_PLIST/ P(NP_max,5), LLIST(NP_max), NP
```
And the hepevt block:
```fortran
      COMMON /HEPEVT/ NEVHEP,NHEP,ISTHEP(NMXHEP),IDHEP(NMXHEP),
     &                JMOHEP(2,NMXHEP),JDAHEP(2,NMXHEP),PHEP(5,NMXHEP),
     &                VHEP(4,NMXHEP)
```
And this loop
```fortran
     DO I=1,NP
         IF (ABS(LLIST(I)).LT.10000) THEN
            ISTHEP(I) = 1
            ICHG(I) = ICHP(ABS(LLIST(I)))
         ELSE
            ISTHEP(I) = 2
            ICHG(I) = 0
         END IF
         IDHEP(I) = ISIB_PID2PDG(LLIST(I))
         PHEP(1,I) = P(I,1)
         PHEP(2,I) = P(I,2)
         PHEP(3,I) = P(I,3)
         PHEP(4,I) = P(I,4)
         PHEP(5,I) = P(I,5)
      END DO
```
The first part of the loop up to IDHEP(I) is required, but the second part is not necessary. PHEP and P both have the type double. PHEP and P are differently ordered in memory, but numpy can easily fix this and present a uniform view.

There are many more cases like this.

I plan to make a sweep over these pieces of interface code to remove unnecessary copies, and I will then adapt the corresponding code in the XEvent classes, which are there to abstract away these differences.

**Comment by afedynitch** [comment]
One of the main reasons for this was the previous hepmc3 function that could deal with the entire momentum and vertex block. IMO, copying a few thousand doubles is not a very high cost wrt to code uniformity, so all Fortran generators produce HEPEVT common blocks. Dunno, it's a stylistic choice. Also, this additional Fortran layer allows for non trivial manipulations to the event stack without dealing with the performance impact and additional copies in python. Imagine, you want to replace some Lambda's with neutrons, correcting for mass shell or so... In Fortran that would be a few lines. 

For Pythia8 something similar must be done anyways since they have a std::vector of structs.

**Comment by HDembinski** [comment]
I agree that code uniformity is important and may outweigh the performance concerns, especially if the cost of copying the doubles is negligible. You may be right that this is not relevant, but have you checked that? You also asked me to measure the performance of the new Event class implementation, so I give that question back to you.

Before I made the most recent changes to pyhepmc, it was indeed important to fill a standard HEPEvt common block, but I have rewritten that code to make it more flexible and since the original HepMC3 implementation is a broken mess. So that limitation does not exist anymore.

> Also, this additional Fortran layer allows for non trivial manipulations to the event stack without dealing with the performance impact and additional copies in python. Imagine, you want to replace some Lambda's with neutrons, correcting for mass shell or so... In Fortran that would be a few lines.

I agree that having the Fortran layer is good in general, we need it at least to have a fast conversion of the internal IDs to PDG ID and to calculate the charge if needed. I am not suggesting to get rid of it entirely. I am just saying that we currently do more than what is necessary AFAIU, but since I may not have the full picture, it is good to discuss this.

**Comment by jncots** [comment]
Solved in #52 and #124
